### PR TITLE
adding Fence client_Creds tests

### PR DESCRIPTION
**Pull request by atharvar28**
PXP-10230

Adding Fence client_credential integration tests

(tested on service versions)
fence - 6.1.0
arborist - master
requestor - master

Ticket to improve the fenceProps functions
https://ctds-planx.atlassian.net/browse/QAT-670

### CODE REVIEW

File: suites/apis/clientCredGrantTest.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+fence-create with --grant-types `client_credentials`
+which return client_id and secret_id
+
+2. run usersync (The client will have been previously added to the Jenkins user.yaml and granted access to
+what it will need for the test, but i think usersync will not grant the access if the client does not
+exist in the Fence DB. So you need to run usersync again during the test, after creating the client)
+
+3. get a client access token using the client ID and secret.
+
+4. hit Fence’s /user endpoint and check the access
```
Comment (paulineribeyre) [inline suites/apis/clientCredGrantTest.js:15]: this step was removed

File: suites/apis/clientCredGrantTest.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  const tokenReq = bash.runCommand(`curl --user "${clientID}:${secretID}" --request POST "https://${process.env.HOSTNAME}/user/oauth2/token?grant_type=client_credentials" -d scope="openid user"`);
+  if (process.env.DEBUG === 'true') {
+    console.log(`data : ${tokenReq}`);
+  }
+  const tokens = JSON.parse(tokenReq);
+  const accessToken = tokens.access_token;
+  console.log(`###Access_token : ${accessToken}`);
+  if (process.env.DEBUG === 'true') {
+    console.log(`###Access_Token : ${accessToken}`);
+  }
```
Comment (paulineribeyre) [inline suites/apis/clientCredGrantTest.js:49]: duplicate logs; let's remove the 1st one?

File: suites/apis/clientCredGrantTest.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  return accessToken;
+}
+
+function deleteClient() {
```
Comment (paulineribeyre) [inline suites/apis/clientCredGrantTest.js:53]: there's already a `deleteClient` function in `fenceProps` - could we move it to `fenceTasks` and reuse it here?

File: suites/apis/clientCredGrantTest.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  const deleteClientCmd = `fence-create client-delete --client ${clientName}`;
+  const deleteClientReq = bash.runCommand(deleteClientCmd, 'fence', takeLastLine);
+  console.log(`Client deleted : ${deleteClientReq}`);
+}
+
+function createClient() {
```
Comment (paulineribeyre) [inline suites/apis/clientCredGrantTest.js:59]: same here; could we move the existing `createClient` function to `fenceTasks` and use it here? it already has a `clientType` parameter that you can use to add logic to make a "client_credentials" client

File: services/apis/requestor/requestorTasks.js [hunk 5]
```diff
@@ -21,6 +21,16 @@ module.exports = {
     }
     return reqID;
   },
+
+  // getting the list of requests in the DB
+  async getRequestList(token) {
+    console.log('Getting list of request for the user');
+    return I.sendGetRequest(
+      `${requestorProps.endpoint.userEndPoint}`,
```
[the remark this answers is not in the document]
Comment (atharvar28) [reply services/apis/requestor/requestorTasks.js:29]: I haven't called this function in the test as of now, as I was waiting for the ticket as well. But what I can do here is, leave the function in current state with change in endpoint. And comment the function call in the test

File: suites/apis/clientCredGrantTest.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  //  create a access request in requestor DB with the clientAccessToken
+  const createRequest = await I.sendPostRequest(
+    `${requestorProps.endpoint.requestEndPoint}`,
+    data,
+    getAccessTokenHeader(clientAccessToken),
+  );
+  if (process.env.DEBUG === 'true') {
+    console.log(createRequest.data);
+    console.log(createRequest.data.request_id);
+  }
```
[the remark this answers is not in the document]
Comment (atharvar28) [reply suites/apis/clientCredGrantTest.js:132]: Lets do this
1. check if the request is successful
2. debug log stmt
3. if request is successful, check if the response data consists of request_id. If expect fails , show the response data

File: run-tests.sh [hunk 7]
```diff
@@ -260,6 +260,9 @@ runTestsIfServiceVersion "@requestor" "requestor" "1.5.0" "2022.02"
 runTestsIfServiceVersion "@requestor" "arborist" "3.2.0" "2021.12"
 runTestsIfServiceVersion "@requestorNew" "requestor" "1.5.1" "2022.06"
 runTestsIfServiceVersion "@requestorNew" "arborist" "3.2.0" "2021.12"
+runTestsIfServiceVersion "@clientCreds" "arborist" "2022.10"
+runTestsIfServiceVersion "@clientCreds" "fence" "6.1.0" "2022.10"
+runTestsIfServiceVersion "@clientCreds" "requestor" "2022.10"
```
Comment (paulineribeyre) [inline run-tests.sh:265]: reminder: before merging the arborist and requestor lines will need to be updated here once the services are updated

Comment (atharvar28) [reply run-tests.sh:265]: 👍

File: run-tests.sh [hunk 8]
```diff
@@ -260,10 +260,10 @@ runTestsIfServiceVersion "@requestor" "requestor" "1.5.0" "2022.02"
 runTestsIfServiceVersion "@requestor" "arborist" "3.2.0" "2021.12"
 runTestsIfServiceVersion "@requestorNew" "requestor" "1.5.1" "2022.06"
 runTestsIfServiceVersion "@requestorNew" "arborist" "3.2.0" "2021.12"
-runTestsIfServiceVersion "@clientCreds" "arborist" "2022.10"
+runTestsIfServiceVersion "@clientCreds" "arborist" "4.0.0" "2022.10"
 runTestsIfServiceVersion "@clientCreds" "fence" "6.1.0" "2022.10"
-runTestsIfServiceVersion "@clientCreds" "requestor" "2022.10"
-
+runTestsIfServiceVersion "@clientCreds" "requestor" "1.8.0" "2022.10"
```
Comment (paulineribeyre) [inline run-tests.sh:265]: 2022.10 is not right

File: services/apis/fence/fenceTasks.js [hunk 9]
```diff
@@ -39,6 +39,47 @@ async function getNoRedirect(url, headers) {
...
 module.exports = {
+  /**
+ * Runs a fence command for creating a client
+ * @param {string} clientName - client name
+ * @param {string} userName - user name
+ * @param {string} clientType - client type (implicit or basic)
+ * @param {string} arboristPolicies - space-delimited list of arborist policies to give to client
+ * @returns {json}
+ */
+  createClient(clientName, userName, clientType, arboristPolicies = null) {
```
[the remark this answers is not in the document]
Comment (atharvar28) [reply services/apis/fence/fenceTasks.js:50]: to refactor the existing fenceTask function for create and delete clients, we would need to change the whole existing logic and also check if the change works for the codes in which these functions are used. So I would leave these functions here and then create new ticket for working on it later

[the remark this answers is not in the document]
Comment (paulineribeyre) [reply services/apis/fence/fenceTasks.js:50]: ok, then i'll approve, could you copy the ticket number here?

**Review by paulineribeyre — DISMISSED** [review]

File: suites/apis/clientCredGrantTest.js [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    console.log(`Request ${I.cache.requestID} deleted successfully`);
+  }
+  // delete the client jenkinsClientTester
+  console.log('Deleting client created for the test ...');
+  fence.do.deleteClient(clientName);
+
+  // revoking the arborist policy for the user
+  console.log('Revoking the arborist policy for the user0 ...');
+  await bash.runCommand(`
+      gen3 devterm curl -X DELETE arborist-service/user/dcf-integration-test-0@planx-pla.net/policy/requestor_integration_test
```
Comment (haraprasadj) [inline suites/apis/clientCredGrantTest.js:72]: can we move this to a separate service (arborist) since this is basically an arborist api call?

Comment (atharvar28) [reply suites/apis/clientCredGrantTest.js:72]: Currently this is the only way  of doing it. Not all the endpoints are exposed via revproxy as [delete endpoint](https://petstore.swagger.io/?url=https://raw.githubusercontent.com/uc-cdis/arborist/master/docs/openapi.yaml#/user/delete_user__username__policy__policyName_) does not have an auth layer. Also the revproxy [conf](https://github.com/uc-cdis/cloud-automation/blob/master/kube/services/revproxy/gen3.nginx.conf/arborist-service.conf) for arborist

File: services/apis/fence/fenceTasks.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (haraprasadj) [reply services/apis/fence/fenceTasks.js:50]: https://ctds-planx.atlassian.net/browse/QAT-670

**Review by haraprasadj — APPROVED** [review]
+1

**Review by paulineribeyre — APPROVED** [review]

**Pull request merged**